File: tests/drivers/python/v1/max_query_length.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from neo4j.v1 import GraphDatabase, basic_auth

driver = GraphDatabase.driver("bolt://localhost:7687",
                              auth=basic_auth("", ""),
                              encrypted=False)

query_template = 'CREATE (n {name:"%s"})'
template_size = len(query_template) - 2  # because of %s
min_len = 1
max_len = 1000000

# binary search because we have to find the maximum size (in number of chars)
# of a query that can be executed via driver
while True:
    assert min_len > 0 and max_len > 0, \
        "The lengths have to be positive values! If this happens something" \
        " is terrible wrong with min & max lengths OR the database" \
        " isn't available."
    property_size = (max_len + min_len) // 2
    try:
        driver.session().run(query_template % ("a" * property_size)).consume()
        if min_len == max_len or property_size + 1 > max_len:
            break
        min_len = property_size + 1
    except Exception as e:
        print("Query size %s is too big!" % (template_size + property_size))
        max_len = property_size - 1

assert property_size == max_len, "max_len probably has to be increased!"

print("\nThe max length of a query from Python driver is: %s\n" %
      (template_size + property_size))

# sessions are not closed bacause all sessions that are
# executed with wrong query size might be broken
driver.close()

File: tests/drivers/java/v5_8/src/main/java/memgraph/MaxQueryLength.java
/**
 * Determines how long could be a query executed
 * from Java driver.
 *
 * Performs binary search until the maximum possible
 * query size has found.
 */
package memgraph;

import static org.neo4j.driver.Values.parameters;

import java.util.*;
import org.neo4j.driver.*;
import org.neo4j.driver.types.*;

public class MaxQueryLength {
  public static void main(String[] args) {
    // init driver
    Config config = Config.builder().withoutEncryption().build();
    Driver driver = GraphDatabase.driver("bolt://localhost:7687", AuthTokens.basic("", ""), config);
    // init query
    int property_size = 0;
    int min_len = 1;
    int max_len = 100000;
    String query_template = "CREATE (n {name:\"%s\"})";
    int template_size = query_template.length() - 2; // because of %s

    // binary search
    while (true) {
      property_size = (max_len + min_len) / 2;
      try (Session session = driver.session()) {
        String property_value = new String(new char[property_size]).replace('\0', 'a');
        String query = String.format(query_template, property_value);
        session.run(query).consume();
        if (min_len == max_len || property_size + 1 > max_len) {
          break;
        }
        min_len = property_size + 1;
      } catch (Exception e) {
        System.out.println(
            String.format("Query length: %d; Error: %s", property_size + template_size, e));
        max_len = property_size - 1;
      }
    }

    // final result
    System.out.println(String.format("\nThe max length of a query executed from "
            + "Java driver is: %s\n",
        property_size + template_size));

    // cleanup
    driver.close();
  }
}
